sphinx/util/parsing.py

```diff
@@ -0,0 +1,97 @@
+
+import contextlib
+from typing import TYPE_CHECKING
+
+from docutils import nodes
+from docutils.statemachine import StringList, string2lines
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from docutils.parsers.rst.states import Inliner, RSTState, Struct
+
+
+def nested_parse_to_nodes(
+    state: RSTState,
+    text: str | StringList,
+    *,
+    source: str = '<generated text>',
+    offset: int = 0,
+    keep_title_context: bool = False,
+) -> list[nodes.Node]:  # Element | nodes.Text
+    """Parse *text* into nodes.
+
+    :param state:
+        The state machine state. Must be a subclass of ``RSTState``.
+    :param text:
+        Text, in string form. ``StringList`` is also accepted.
+    :param source:
+        The text's source, used when creating a new ``StringList``.
+    :param offset:
+        The offset of the content.
+    :param keep_title_context:
+        If this is False (the default), then *content* is parsed as if it were
+        an independent document, meaning that title decorations (e.g. underlines)
+        do not need to match the surrounding document.
+        This is useful when the parsed content comes from
+        a completely different context, such as docstrings.
+        If this is True, then title underlines must match those in
+        the surrounding document, otherwise errors will occur. TODO: check!
+    """
+    document = state.document
+    content = _text_to_string_list(
+        text, source=source, tab_width=document.settings.tab_width,
+    )
+    node = nodes.Element()  # Anonymous container for parsing
+    node.document = document
+
+    if keep_title_context:
+        state.nested_parse(content, offset, node, match_titles=True)
+    else:
+        with _fresh_title_style_context(state):
+            state.nested_parse(content, offset, node, match_titles=True)
+    return node.children
+
+
+@contextlib.contextmanager
+def _fresh_title_style_context(state: RSTState) -> Iterator[None]:
+    # hack around title style bookkeeping
+    memo = state.memo
+    surrounding_title_styles: list[str | tuple[str, str]] = memo.title_styles
+    surrounding_section_level: int = memo.section_level
+    # clear current title styles
+    memo.title_styles = []
+    memo.section_level = 0
+    try:
+        yield
+    finally:
+        # reset title styles
+        memo.title_styles = surrounding_title_styles
+        memo.section_level = surrounding_section_level
+
+
+def inliner_parse_text(
+    text: str, *, state: RSTState, lineno: int = 1,
+) -> tuple[list[nodes.Node], list[nodes.system_message]]:
+    """Parse *text* as inline nodes.
+
+    The text cannot contain any structural elements (headings, transitions,
+    directives, etc), so should be a simple line or paragraph of text.
+    """
+    inliner: Inliner = state.inliner
+    memo: Struct = state.memo
+    parent: nodes.Element = state.parent
+    return inliner.parse(text, lineno, memo, parent)
```

Comment:
why not just use `state.inline_text(text, lineno)` here, as per the code you are replacing?

Its not a hard fix, but this definitely breaks the current implementation of myst-parser

Reply:
Ahh, I didn't realise -- reverted to a simpler implementation in #12504.

A